### PR TITLE
Message and MultiChats link to backend

### PR DESCRIPTION
**Pull request by Lingtao-Cai**
Connected Messaging and MultiChat with backend

1. Implement creates/close conversation, any new conversation can be created by calling createConversation function.
2. Implement functionality of messaging in chats.
3. Implement showing profile pictures on both MultiChats and MessagingList pages.
4. Update the date format, it now shows the correct time in messaging and chats.
5. add a converter in MessagingActivity for converting ISO Date (8601) to a date object and split it into a String[] (see comments on this function for more details).

### CODE REVIEW

**Review by Manoutofpower — APPROVED** [review]

File: app/src/main/java/com/example/ar_reshare/BackendController.java [hunk 1]
```diff
@@ -273,6 +288,253 @@ public void onFailure(Call<ResponseBody> call, Throwable t) {
...
 
+
+    public static boolean createConversation(Integer listingID, BackendCallback callback){
+        Retrofit retrofit = new Retrofit.Builder()
+                .baseUrl(URL)
+                .addConverterFactory(GsonConverterFactory.create())
+                .build();
+
+        String bodyString =
+                String.format("{\n  \"listingID\": \"%d\"\n}",
```
Comment (artur-varosyan) [inline app/src/main/java/com/example/ar_reshare/BackendController.java:299]: Hi Lingtao, I was wondering if you could please replace this, with the method @HellinSun found. That is, using a _JsonObject_. It is a bit more readable and neater then! 😄

File: app/src/main/java/com/example/ar_reshare/BackendController.java [hunk 2]
```diff
@@ -273,6 +288,253 @@ public void onFailure(Call<ResponseBody> call, Throwable t) {
...
+
+    public static boolean getConversationDescriptors(ChatBackendCallback callback) {
+        Retrofit retrofit = new Retrofit.Builder()
+                .baseUrl(URL)
+                .addConverterFactory(GsonConverterFactory.create())
+                .build();
+
+//        String bodyString =
+//                String.format("{\n  \"name\": \"%d\"\n}",
+//                        listingID);
```
Comment (artur-varosyan) [inline app/src/main/java/com/example/ar_reshare/BackendController.java:359]: Can you please remove the redundant commented lines of code?

File: app/src/main/java/com/example/ar_reshare/BackendController.java [hunk 3]
```diff
@@ -273,6 +288,253 @@ public void onFailure(Call<ResponseBody> call, Throwable t) {
...
+
+    public static boolean sendConversationMessage(Integer conversationID, String textContent, String mediaContent, BackendCallback callback) throws JSONException {
+        Retrofit retrofit = new Retrofit.Builder()
+                .baseUrl(URL)
+                .addConverterFactory(GsonConverterFactory.create())
+                .build();
+
+        JSONObject jsonObject = new JSONObject();
+        jsonObject.put("conversationID", conversationID);
+        jsonObject.put("textContent", textContent);
```
Comment (artur-varosyan) [inline app/src/main/java/com/example/ar_reshare/BackendController.java:447]: This method is great!

File: app/src/main/java/com/example/ar_reshare/MessageListAdapter.java [hunk 4]
```diff
@@ -22,12 +25,16 @@ public class MessageListAdapter extends RecyclerView.Adapter {
 
     private Context mContext;
     private List<Message> mMessageList;
+    private Message.MessageResult messageResult;
+    int loggedInUserID;
```
Comment (artur-varosyan) [inline app/src/main/java/com/example/ar_reshare/MessageListAdapter.java:29]: Could you please add a _public_ access modifier here? Sorry, this is my fault I should have done it in my last pull request!

File: app/src/main/java/com/example/ar_reshare/SplashActivity.java [hunk 5]
```diff
@@ -59,7 +59,7 @@ private void proceed() {
         timer.schedule(new TimerTask() {
             @Override
             public void run() {
-                Intent intent = new Intent(SplashActivity.this, ARActivity.class);
+                Intent intent = new Intent(SplashActivity.this, FallbackActivity.class);
```
Comment (artur-varosyan) [inline app/src/main/java/com/example/ar_reshare/SplashActivity.java:62]: I do not think this should have been committed Lingtao! I am assuming you did it just for testing.

**Review by artur-varosyan — CHANGES_REQUESTED** [review]
Hi Lingtao. You have done a great job - so much functionality added and so many backend endpoints implemented. Also, the code quality and structure are really good. I just requested a few changes. Let me know if you disagree or have any questions.

File: app/src/main/java/com/example/ar_reshare/MessageListAdapter.java (hunk 4 above)
Comment (artur-varosyan) [reply app/src/main/java/com/example/ar_reshare/MessageListAdapter.java:29]: Actually, my mistake this is the wrong file. Feel free to leave it with the default or change it to private. (resolved)

**Review by HellinSun — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/ar_reshare/BackendController.java (hunk 1 above)
Comment (Lingtao-Cai) [reply app/src/main/java/com/example/ar_reshare/BackendController.java]: I have now replaced all body string creating with JSON object related methods, Thanks Artur!

File: app/src/main/java/com/example/ar_reshare/BackendController.java (hunk 3 above)
Comment (Lingtao-Cai) [reply app/src/main/java/com/example/ar_reshare/BackendController.java:437]: 😄

File: app/src/main/java/com/example/ar_reshare/SplashActivity.java (hunk 5 above)
Comment (Lingtao-Cai) [reply app/src/main/java/com/example/ar_reshare/SplashActivity.java]: Yeah, it is changed for testing due to the simulator's AR, I have now fixed it!

File: app/src/main/java/com/example/ar_reshare/Message.java [hunk 6]
```diff
@@ -1,28 +1,131 @@
...
     private String imageUrl;
+    private Bitmap profileIcon;
 
-    public Message(String message, User sender, String createdTime) {
+    public Message(Integer senderID, String message, String createdTime, String mediaContentMimetype, String imageUrl) {
+        this.senderID = senderID;
         this.message = message;
-        this.sender = sender;
         this.createdTime = createdTime;
+       // this.mediaContentMimetype = mediaContentMimetype;
```
[the remark this answers is not in the document]
Comment (Lingtao-Cai) [reply app/src/main/java/com/example/ar_reshare/Message.java]: done!

File: app/src/main/java/com/example/ar_reshare/BackendController.java [hunk 7]
```diff
@@ -321,7 +590,11 @@ public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response)
                 @Override
                 public void onFailure(Call<ResponseBody> call, Throwable t) {
                     System.out.println("Failure");
+
+                    callback.onBackendResult(false, "Failed to create new conversation");
```
[the remark this answers is not in the document]
Comment (Lingtao-Cai) [reply app/src/main/java/com/example/ar_reshare/BackendController.java]: This maybe come from the auto-merge and has been ignored when I solved conflicts, it now should be fixed!

File: app/src/main/java/com/example/ar_reshare/BackendController.java [hunk 8]
```diff
@@ -300,10 +562,17 @@ public static boolean addProduct(String title, String description, String countr
         RequestBody body = RequestBody.create(MediaType.parse("application/json"), bodyString);
         BackendService service = retrofit.create(BackendService.class);
         Call<ResponseBody> call = service.addProduct(JWT, body);
+
         try {
             call.enqueue(new Callback<ResponseBody>() {
                 @Override
                 public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
+                    try {
+                        String string = response.body().string();
```
[the remark this answers is not in the document]
Comment (Lingtao-Cai) [reply app/src/main/java/com/example/ar_reshare/BackendController.java]: Done!

**Review by artur-varosyan — APPROVED** [review]
Looks great now thanks Lingtao!

**Review by arfi2212 — APPROVED** [review]
Brilliant stuff Lingtao! Very clean code too 😄

**Review by HellinSun — APPROVED** [review]
So much work done!!! Great job!

**Pull request merged**